File: osint_nexus/core/db/result_repository.py
from __future__ import annotations

import logging
from typing import cast

from osint_nexus.core.db.base import DatabaseEngine
from osint_nexus.core.type_defs import JSONObject, JSONValue

logger = logging.getLogger("osint_nexus.db.result_repository")


class ResultRepository:
    def __init__(self, engine: DatabaseEngine) -> None:
        self.engine = engine

    async def save(self, username: str, platform: str, found: bool) -> None:
        try:
            await self.engine.execute(
                "INSERT INTO results (username, platform, found) VALUES (?, ?, ?)",
                (username, platform, int(found)),
            )
        except Exception as exc:
            logger.error("Failed to save result: %s", exc, exc_info=True)

    async def save_batch(self, query: str, data: list[tuple[JSONValue, ...]]) -> None:
        try:
            await self.engine.executemany(query, data)
        except Exception as exc:
            logger.error("Failed to perform batch insert: %s", exc, exc_info=True)

    async def query(
        self,
        username: str | None = None,
        platform: str | None = None,
        limit: int = 100,
    ) -> list[JSONObject]:
        query = "SELECT id, username, platform, found, timestamp FROM results WHERE 1=1"
        params: list[JSONValue] = []
        if username:
            query += " AND username = ?"
            params.append(username)
        if platform:
            query += " AND platform = ?"
            params.append(platform)
        query += " ORDER BY timestamp DESC LIMIT ?"
        params.append(limit)

        try:
            rows = await self.engine.fetchall(query, tuple(params))
            return cast(list[JSONObject], rows)
        except Exception as exc:
            logger.error("Query failed: %s", exc, exc_info=True)
            return []

File: internal/db/result_repository.go
package db

import (
	"fmt"
	"github.com/osint-nexus/internal/types"
)

// ResultRepository handles CRUD operations for scan results.
type ResultRepository struct {
	engine *SQLiteEngine
}

// NewResultRepository initializes a new ResultRepository.
func NewResultRepository(engine *SQLiteEngine) *ResultRepository {
	return &ResultRepository{engine: engine}
}

// Save inserts a new scan result.
func (r *ResultRepository) Save(username, platform string, found bool) error {
	foundInt := 0
	if found {
		foundInt = 1
	}
	query := "INSERT INTO results (username, platform, found) VALUES (?, ?, ?)"
	return r.engine.Exec(query, username, platform, foundInt)
}

// SaveBatch inserts multiple scan results in a single transaction.
func (r *ResultRepository) SaveBatch(query string, data ...[]interface{}) error {
	return r.engine.ExecMany(query, data...)
}

// Query retrieves scan results based on filters.
func (r *ResultRepository) Query(username, platform string, limit int) ([]types.ScanResult, error) {
	query := "SELECT id, username, platform, found, timestamp FROM results WHERE 1=1"
	var params []interface{}

	if username != "" {
		query += " AND username = ?"
		params = append(params, username)
	}
	if platform != "" {
		query += " AND platform = ?"
		params = append(params, platform)
	}
	query += " ORDER BY timestamp DESC LIMIT ?"
	params = append(params, limit)

	rows, err := r.engine.DB().Query(query, params...)
	if err != nil {
		return nil, fmt.Errorf("db: query: %w", err)
	}
	defer rows.Close()

	var results []types.ScanResult
	for rows.Next() {
		var res types.ScanResult
		var foundInt int
		if err := rows.Scan(&res.ID, &res.Username, &res.Platform, &foundInt, &res.Timestamp); err != nil {
			return nil, fmt.Errorf("db: scan: %w", err)
		}
		res.Found = (foundInt == 1)
		results = append(results, res)
	}
	return results, nil
}
